crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,268 @@
+use wasm_bindgen::throw_str;
+
+pub mod callbacks {
+    use crate::{
+        blob::Blob,
+        file_reader::{FileReadError, ReadyState},
+    };
+    use gloo_events::EventListener;
+    use std::sync::{Arc, Mutex};
+    use wasm_bindgen::{throw_str, JsCast, UnwrapThrowExt};
+
+    /// A guard object that aborts the file read when dropped (if the read isn't already
+    /// finished).
+    #[derive(Debug)]
+    pub struct FileReader {
+        reader: web_sys::FileReader,
+        load_listener: EventListener,
+        error_listener: EventListener,
+    }
+
+    impl std::ops::Drop for FileReader {
+        fn drop(&mut self) {
+            if !ReadyState::from(self.reader.ready_state()).is_done() {
+                self.reader.abort();
+            }
+        }
+    }
+
+    /// Read the contents of `blob` into a text string, and then run the callback.
+    ///
+    /// If the returned `FileReader` is dropped before the callback is invoked, the read will be
+    /// cancelled.
+    pub fn read_as_text<F>(blob: &Blob, callback: F) -> FileReader
+    where
+        F: FnOnce(Result<String, FileReadError>) + 'static,
+    {
+        // we need to be able to run the FnOnce, while proving to the compiler that it can only run
+        // once. The easiest way is to `take` it out of an Option. The `Arc` and `Mutex` are
+        // because we need shared ownership and mutability (for the `take`).
+        let load_callback: Arc<
+            Mutex<Option<Box<dyn FnOnce(Result<String, FileReadError>) + 'static>>>,
+        > = Arc::new(Mutex::new(Some(Box::new(callback))));
+        let error_callback = load_callback.clone();
+
+        let reader = web_sys::FileReader::new().unwrap_throw();
+
+        // Either the load listener or the error listener will be called, never both (so FnOnce is
+        // ok).
+        let load_listener = EventListener::new(&reader, "load", move |event| {
+            let reader = reader_from_event(&event);
+            let result = reader.result().unwrap_throw().as_string().unwrap_throw();
+            load_callback.lock().unwrap_throw().take().unwrap_throw()(Ok(result));
+        });
+
+        let error_listener = EventListener::new(&reader, "error", move |event| {
+            let error = error_from_event(&event);
+            error_callback.lock().unwrap_throw().take().unwrap_throw()(Err(error));
+        });
+
+        reader.read_as_text(blob.as_ref()).unwrap_throw();
```

Comment:
Nit: there should be a newline here, just for visual spacing.

```suggestion
        reader.read_as_text(blob.as_ref()).unwrap_throw();

```